### PR TITLE
v0.0.5

### PR DESCRIPTION
**Pull request by sol-znn**
## v0.0.5

This branch adds support for the following:
- Refactored to align with the [znn_cli_csharp](https://github.com/hypercore-one/znn_cli_csharp)
- Added support for HTLC, Spork, Bridge, Liquidity contracts
- Added admin-only functionality (Bridge/Liquidity)
- Added ability to donate to AZ and query Stats RPC
- Added `chainId` option (default, auto, manual)
- Updated menu, including an admin-only menu
   - Updated `verbose` flag to display a brief description for commands
   - Updated `help` flag to display all options for a category
- Updated Makefile: bugfix + support for Windows

**Limitations**
- Some [orchestrator-specific](https://github.com/hypercore-one/znn_cli_dart/blob/feature/v0.0.5/lib/orchestrator.dart) functions were not included due to low value/increased complexity.  
These can be omitted since they're handled by orchestrators.

### CODE REVIEW

File: src/general.dart [hunk 1]
```diff
@@ -0,0 +1,326 @@
+import 'dart:convert';
+import 'package:convert/convert.dart';
+import 'package:dcli/dcli.dart' hide verbose;
+import 'package:znn_sdk_dart/znn_sdk_dart.dart';
+import 'src.dart';
+
+void generalMenu() {
+  print('  ${white('General')}');
+  print(
+      '    send toAddress amount [${green('ZNN')}/${blue('QSR')}/${magenta('ZTS')}]');
```
Comment (KingGorrin) [inline src/general.dart]: The optional message argument is missing.

File: src/general.dart [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+  print('    receive blockHash');
+  print('    receiveAll');
+  print('    unreceived');
+  print('    unconfirmed');
+  print('    balance address');
+  print('    frontierMomentum');
+  print('    createHash "string" [hashType preimageLength]');
```
Comment (KingGorrin) [inline src/general.dart]: `createHash` only has two arguments `[hashType preimageLength]`

File: src/general.dart [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+      return;
+
+    default:
+      invalidCommand();
+  }
+}
+
+void _version() {
+  print('$znnCli v$znnCliVersion using Zenon SDK v$znnSdkVersion');
+  print(getZnndVersion(znnDaemon));
```
Comment (KingGorrin) [inline src/general.dart]: Why show the znnd information?

File: src/general.dart [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+        'send toAddress amount [${green('ZNN')}/${blue('QSR')}/${magenta('ZTS')}]');
+    return;
+  }
+
+  Address recipient = Address.parse(args[1]);
+  TokenStandard tokenStandard = getTokenStandard(args[3]);
+  Token token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
+  BigInt amount =
+      AmountUtils.extractDecimals(num.parse(args[2]), token.decimals);
+
```
Comment (KingGorrin) [inline src/general.dart]: Negative or zero amounts are not checked.

File: src/general.dart [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+  if (!await hasBalance(znnClient, address, tokenStandard, amount)) {
+    return;
+  }
+
+  var block = AccountBlockTemplate.send(recipient, tokenStandard, amount);
+
+  if (args.length == 5) {
+    block.data = AsciiEncoder().convert(args[4]);
```
Comment (KingGorrin) [inline src/general.dart]: Can UFT8 encoding be used instead?

File: src/general.dart [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+    print(
+        '${yellow('Warning!')} preimageLength is less than $htlcPreimageDefaultLength and may be insecure');
+  }
+  preimage = generatePreimage(preimageLength);
+  print('Preimage: ${hex.encode(preimage)}');
+
+  switch (hashType) {
+    case 1:
+      hash = Hash.fromBytes(await Crypto.sha256Bytes(preimage));
+      print('SHA-256 Hash: $hash');
```
Comment (KingGorrin) [inline src/general.dart]: SHA2-256

File: src/general.dart [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+      return;
+    default:
+      hash = Hash.digest(preimage);
+      print('SHA-3 Hash: $hash');
```
Comment (KingGorrin) [inline src/general.dart]: SHA3-256

File: src/htlc.dart [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+  late Hash hashLock;
+  int keyMaxSize = htlcPreimageMaxLength;
+  int hashType = 0;
+  List<int> preimage = generatePreimage(htlcPreimageDefaultLength);
+
+  int htlcTimelockMinHours = 1;
+  int htlcTimelockMaxHours = htlcTimelockMinHours * 24;
+
+  try {
+    hashLockedAddress = Address.parse(args[1]);
```
Comment (KingGorrin) [inline src/htlc.dart]: No checks are made to prevent empty or embedded addresses.

File: src/htlc.dart [hunk 9]
```diff
@@ -0,0 +1,657 @@
...
+  } catch (e) {
+    print('${red('Error!')} hashLockedAddress must be a valid address');
+    return;
+  }
+
+  Token token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
+  amount = AmountUtils.extractDecimals(num.parse(args[3]), token.decimals);
```
Comment (KingGorrin) [inline src/htlc.dart]: Amount as string?

File: src/htlc.dart [hunk 10]
```diff
@@ -0,0 +1,657 @@
...
+      hashLock = Hash.parse(args[6]);
+    } catch (e) {
+      print('${red('Error!')} hashLock is not a valid hash');
+      return;
+    }
+  } else {
+    switch (hashType) {
+      case 1:
+        hashLock = Hash.fromBytes(await Crypto.sha256Bytes(preimage));
+        return;
```
Comment (KingGorrin) [inline src/htlc.dart]: I think this needs to be a `break`?

File: src/htlc.dart [hunk 11]
```diff
@@ -0,0 +1,657 @@
...
+  }
+
+  Token token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
+  amount = AmountUtils.extractDecimals(num.parse(args[3]), token.decimals);
+
+  if (amount <= BigInt.zero) {
+    print('${red('Error!')} amount must be greater than 0');
+    return;
+  }
+
```
Comment (KingGorrin) [inline src/htlc.dart]: No balance check is made. `if (!(await hasBalance()) return;`

File: src/htlc.dart [hunk 12]
```diff
@@ -0,0 +1,657 @@
...
+}
+
+Future<void> _inspect() async {
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('htlc.inspect blockHash');
+    return;
+  }
+
+  Hash blockHash = Hash.parse(args[1]);
```
Comment (KingGorrin) [inline src/htlc.dart]: catch Hash.parse exception.

File: src/htlc.dart [hunk 13]
```diff
@@ -0,0 +1,657 @@
...
+  }
+
+  if (preimage.isEmpty) {
+    print('${red('Error!')} Cannot unlock htlc. Invalid pre-image');
+    return;
+  }
+
+  switch (hashType) {
+    case 1:
+      print('HashType 1 detected. Encoding preimage to SHA-256...');
```
Comment (KingGorrin) [inline src/htlc.dart]: SHA2-256

File: src/accelerator.dart [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+  if (tokenStandard != znnZts || tokenStandard != qsrZts) {
+    print(
+        '${red('Error!')} You can only send ${green('ZNN')} or ${blue('QSR')}.');
+    return;
+  }
+
+  Token token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
+  BigInt amount =
+      AmountUtils.extractDecimals(num.parse(args[1]), token.decimals);
+  if (amount == BigInt.zero) {
```
Comment (KingGorrin) [inline src/accelerator.dart]: Check if amount is bigger than 0. `if (amount <= BigInt.zero)`

File: src/accelerator.dart [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    print('${red('Error!')} You cannot send that amount.');
+    return;
+  }
+
+  if (!await hasBalance(znnClient, address, tokenStandard, amount)) {
+    return;
+  }
+
+  print(
+      'Donating ${AmountUtils.addDecimals(amount, token.decimals)} ${token.symbol} to Accelerator Z ...');
```
Comment (KingGorrin) [inline src/accelerator.dart]: Accelerator-Z

File: src/pillar.dart [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+  if (depositedQsr < qsrAmount) {
+    print(
+        'Depositing ${AmountUtils.addDecimals(qsrAmount - depositedQsr, coinDecimals)} ${blue('QSR')} for the Pillar registration');
+    await znnClient
+        .send(znnClient.embedded.pillar.depositQsr(qsrAmount - depositedQsr));
+  }
+  print('Registering Pillar ...');
+  await znnClient.send(znnClient.embedded.pillar.register(
+      newName,
+      Address.parse(args[2]),
```
Comment (KingGorrin) [inline src/pillar.dart]: Parse and check empty or embedded.

File: src/plasma.dart [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+      ' plasma with ${AmountUtils.addDecimals(plasmaInfo.qsrAmount, coinDecimals)} ${blue('QSR')} fused.');
+}
+
+Future<void> _fuse() async {
+  if (args.length != 3) {
+    print('Incorrect number of arguments. Expected:');
+    print('plasma.fuse toAddress amount (in ${blue('QSR')})');
+    return;
+  }
+  Address beneficiary = Address.parse(args[1]);
```
Comment (KingGorrin) [inline src/plasma.dart]: Check parse and empty or embedded address.

File: src/plasma.dart [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+  print('Done');
+}
+
+Future<void> _cancel() async {
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('plasma.cancel id');
+    return;
+  }
+  Hash id = Hash.parse(args[1]);
```
Comment (KingGorrin) [inline src/plasma.dart]: Try/catch parse

File: src/staking.dart [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+  print('Done');
+}
+
+Future<void> _revoke() async {
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('stake.revoke id');
+    return;
+  }
+  Hash hash = Hash.parse(args[1]);
```
Comment (KingGorrin) [inline src/staking.dart]: Try/catch hash

File: src/token.dart [hunk 20]
```diff
@@ -0,0 +1,353 @@
...
+  }
+}
+
+Future<void> _getByStandard() async {
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('token.getByStandard tokenStandard');
+    return;
+  }
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
```
Comment (KingGorrin) [inline src/token.dart]: Use GetTokenStandard()

File: src/token.dart [hunk 21]
```diff
@@ -0,0 +1,353 @@
...
+  Token token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
```
Comment (KingGorrin) [inline src/token.dart]: Check if token exist

File: src/token.dart [hunk 22]
```diff
@@ -0,0 +1,353 @@
...
+}
+
+Future<void> _getByOwner() async {
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('token.getByOwner ownerAddress');
+    return;
+  }
+  String type = 'Token';
+  Address ownerAddress = Address.parse(args[1]);
```
Comment (KingGorrin) [inline src/token.dart]: Check parse address.

File: src/token.dart [hunk 23]
```diff
@@ -0,0 +1,353 @@
...
+}
+
+Future<void> _transferOwnership() async {
+  if (args.length != 3) {
+    print('Incorrect number of arguments. Expected:');
+    print('token.transferOwnership tokenStandard newOwnerAddress');
+    return;
+  }
+  print('Transferring ZTS token ownership ...');
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
```
Comment (KingGorrin) [inline src/token.dart]: Use GetTokenStandard()

File: src/token.dart [hunk 24]
```diff
@@ -0,0 +1,353 @@
...
+  print('Done');
+}
+
+Future<void> _mint() async {
+  if (args.length != 4) {
+    print('Incorrect number of arguments. Expected:');
+    print('token.mint tokenStandard amount receiveAddress');
+    return;
+  }
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
```
Comment (KingGorrin) [inline src/token.dart]: Use GetTokenStandard()

File: src/token.dart [hunk 25]
```diff
@@ -0,0 +1,353 @@
...
+
+Future<void> _transferOwnership() async {
+  if (args.length != 3) {
+    print('Incorrect number of arguments. Expected:');
+    print('token.transferOwnership tokenStandard newOwnerAddress');
+    return;
+  }
+  print('Transferring ZTS token ownership ...');
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
+  Address newOwnerAddress = Address.parse(args[2]);
```
Comment (KingGorrin) [inline src/token.dart]: Check parse address

File: src/token.dart [hunk 26]
```diff
@@ -0,0 +1,353 @@
...
+}
+
+Future<void> _disableMint() async {
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('token.disableMint tokenStandard');
+    return;
+  }
+  print('Disabling ZTS token mintable flag ...');
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
```
Comment (KingGorrin) [inline src/token.dart]: Use GetTokenStandard()

File: src/token.dart [hunk 27]
```diff
@@ -0,0 +1,353 @@
...
+  var token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
```
Comment (KingGorrin) [inline src/token.dart]: Check if token exists

File: src/token.dart [hunk 28]
```diff
@@ -0,0 +1,353 @@
...
+Future<void> _transferOwnership() async {
+  if (args.length != 3) {
+    print('Incorrect number of arguments. Expected:');
+    print('token.transferOwnership tokenStandard newOwnerAddress');
+    return;
+  }
+  print('Transferring ZTS token ownership ...');
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
+  Address newOwnerAddress = Address.parse(args[2]);
+  var token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
```
Comment (KingGorrin) [inline src/token.dart]: Check if token exists

File: src/liquidity.dart [hunk 29]
```diff
@@ -0,0 +1,625 @@
...
+}
+
+Future<void> _getRewardTotal() async {
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('liquidity.getRewardTotal address');
+    return;
+  }
+
+  Address address = Address.parse(args[1]);
```
Comment (KingGorrin) [inline src/liquidity.dart]: Try/Catch parse address

File: src/liquidity.dart [hunk 30]
```diff
@@ -0,0 +1,625 @@
...
+}
+
+Future<void> _getStakeEntries() async {
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('liquidity.getStakeEntries address');
+    return;
+  }
+
+  Address address = Address.parse(args[1]);
```
Comment (KingGorrin) [inline src/liquidity.dart]: Try/catch parse address

File: src/liquidity.dart [hunk 31]
```diff
@@ -0,0 +1,625 @@
...
+}
+
+Future<void> _getUncollectedReward() async {
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('liquidity.getUncollectedReward address');
+    return;
+  }
+
+  Address address = Address.parse(args[1]);
```
Comment (KingGorrin) [inline src/liquidity.dart]: Try/catch parse address

File: src/liquidity.dart [hunk 32]
```diff
@@ -0,0 +1,625 @@
...
+Future<void> _stake() async {
+  if (args.length != 4) {
+    print('Incorrect number of arguments. Expected:');
+    print('liquidity.stake duration (in months) amount tokenStandard');
+    return;
+  }
+
+  int months = int.parse(args[1]);
+  int duration = months * 24 * 60 * 60;
+  TokenStandard tokenStandard = TokenStandard.parse(args[3]);
```
Comment (KingGorrin) [inline src/liquidity.dart]: Use GetTokenStandard

File: src/liquidity.dart [hunk 33]
```diff
@@ -0,0 +1,625 @@
...
+}
+
+Future<void> _cancelStake() async {
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('liquidity.cancelStake id');
+    return;
+  }
+
+  Hash id = Hash.parse(args[1]);
```
Comment (KingGorrin) [inline src/liquidity.dart]: Try/catch parse hash

File: src/liquidity.dart [hunk 34]
```diff
@@ -0,0 +1,625 @@
...
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('liquidity.guardian.proposeAdmin address');
+    return;
+  }
+
+  String currentAdmin = (await znnClient.embedded.liquidity.getLiquidityInfo())
+      .administrator
+      .toString();
+  Address newAdmin = Address.parse(args[1]);
```
Comment (KingGorrin) [inline src/liquidity.dart]: Try/catch parse address and check new admin must be an user address.

File: src/liquidity.dart [hunk 35]
```diff
@@ -0,0 +1,625 @@
...
+Future<void> _changeAdmin() async {
+  _isAdmin();
+
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('liquidity.admin.changeAdmin address');
+    return;
+  }
+
+  Address newAdmin = Address.parse(args[1]);
```
Comment (KingGorrin) [inline src/liquidity.dart]: Try/catch parse address and check not empty or embedded

File: src/liquidity.dart [hunk 36]
```diff
@@ -0,0 +1,625 @@
...
+      Address guardian = Address.parse(args[i]);
+      if (guardian != emptyAddress) {
+        guardians.add(guardian);
+      }
+    } catch (e) {
+      print('${red('Error!')} ${args[i]} is not a valid address');
+      return;
+    }
+  }
+
```
Comment (KingGorrin) [inline src/liquidity.dart]: Nomination works with time challenges, see  C# impl. for details.

File: src/liquidity.dart [hunk 37]
```diff
@@ -0,0 +1,625 @@
...
+    print('liquidity.admin.nominateGuardians address1 address2 ... addressN');
+    return;
+  }
+
+  List<Address> guardians = [];
+
+  for (int i = 1; i < args.length; i++) {
+    try {
+      Address guardian = Address.parse(args[i]);
+      if (guardian != emptyAddress) {
```
Comment (KingGorrin) [inline src/liquidity.dart]: Check not empty or embedded and all are distinct

File: src/liquidity.dart [hunk 38]
```diff
@@ -0,0 +1,625 @@
...
+  AccountBlockTemplate nominateGuardians =
+      znnClient.embedded.liquidity.nominateGuardians(guardians);
+  nominateGuardians = await znnClient.send(nominateGuardians);
+  print('Done');
+}
+
+Future<void> _unlockStakeEntries() async {
+  AccountBlockTemplate unlockLiquidityStakeEntries =
+      znnClient.embedded.liquidity.unlockLiquidityStakeEntries(
+          TokenStandard.parse('zts17d6yr02kh0r9qr566p7tg6'));
```
Comment (KingGorrin) [inline src/liquidity.dart]: Why is the token standard hardcoded?

File: src/plasma.dart [hunk 39]
```diff
@@ -0,0 +1,154 @@
...
+    print('Incorrect number of arguments. Expected:');
+    print('plasma.list [pageIndex pageSize]');
+    return;
+  }
+  int pageIndex = 0;
+  int pageSize = 25;
+  if (args.length == 3) {
+    pageIndex = int.parse(args[1]);
+    pageSize = int.parse(args[2]);
+  }
```
Comment (KingGorrin) [inline src/plasma.dart]: Assert pageIndex/pageSize constraints

File: src/spork.dart [hunk 40]
```diff
@@ -0,0 +1,106 @@
...
+    print('Incorrect number of arguments. Expected:');
+    print('spork.list [pageIndex pageSize]');
+    return;
+  }
+  int pageIndex = 0;
+  int pageSize = rpcMaxPageSize;
+  if (args.length == 3) {
+    pageIndex = int.parse(args[1]);
+    pageSize = int.parse(args[2]);
+  }
```
Comment (KingGorrin) [inline src/spork.dart]: Assert pageIndex/pageSize constraints

File: src/staking.dart [hunk 41]
```diff
@@ -0,0 +1,155 @@
...
+    print('Incorrect number of arguments. Expected:');
+    print(' stake.list [pageIndex pageSize]');
+    return;
+  }
+  int pageIndex = 0;
+  int pageSize = 25;
+  if (args.length == 3) {
+    pageIndex = int.parse(args[1]);
+    pageSize = int.parse(args[2]);
+  }
```
Comment (KingGorrin) [inline src/staking.dart]: Assert pageIdex/pageSize constraints

File: src/token.dart [hunk 42]
```diff
@@ -0,0 +1,353 @@
...
+    print('Incorrect number of arguments. Expected:');
+    print('token.list [pageIndex pageSize]');
+    return;
+  }
+  int pageIndex = 0;
+  int pageSize = 25;
+  if (args.length == 3) {
+    pageIndex = int.parse(args[1]);
+    pageSize = int.parse(args[2]);
+  }
```
Comment (KingGorrin) [inline src/token.dart]: Assert pageIndex/pageSize constraints

File: src/token.dart [hunk 43]
```diff
@@ -0,0 +1,353 @@
...
+
+Future<void> _mint() async {
+  if (args.length != 4) {
+    print('Incorrect number of arguments. Expected:');
+    print('token.mint tokenStandard amount receiveAddress');
+    return;
+  }
+  TokenStandard tokenStandard = TokenStandard.parse(args[1]);
+  BigInt amount = BigInt.parse(args[2]);
+  Address mintAddress = Address.parse(args[3]);
```
Comment (KingGorrin) [inline src/token.dart]: Try/catch parse address

File: src/bridge.dart [hunk 44]
```diff
@@ -0,0 +1,1197 @@
...
+
+  int networkClass = int.parse(args[1]);
+  int chainId = int.parse(args[2]);
+  String toAddress = args[3]; // must be EVM-compatible
+  TokenStandard tokenStandard = getTokenStandard(args[5]);
+  Token token = (await znnClient.embedded.token.getByZts(tokenStandard))!;
+  BigInt amount =
+      AmountUtils.extractDecimals(num.parse(args[4]), token.decimals);
+
+  if (amount == BigInt.zero) {
```
Comment (KingGorrin) [inline src/bridge.dart]: amount must be greater than 0

File: src/bridge.dart [hunk 45]
```diff
@@ -0,0 +1,1197 @@
...
+}
+
+Future<void> _wrapGet() async {
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('bridge.wrap.get id');
+    return;
+  }
+
+  Hash id = Hash.parse(args[1]);
```
Comment (KingGorrin) [inline src/bridge.dart]: Try/catch parse hash

File: src/bridge.dart [hunk 46]
```diff
@@ -0,0 +1,1197 @@
...
+}
+
+Future<void> _unwrapRedeem() async {
+  if (args.length != 3) {
+    print('Incorrect number of arguments. Expected:');
+    print('bridge.unwrap.redeem transactionHash logIndex');
+    return;
+  }
+
+  Hash transactionHash = Hash.parse(args[1]);
```
Comment (KingGorrin) [inline src/bridge.dart]: Try/catch parse hash

File: src/bridge.dart [hunk 47]
```diff
@@ -0,0 +1,1197 @@
...
+    decimals = token.decimals;
+  }
+
+  print('Id: ${request.transactionHash}');
+  print('   Network Class: ${request.networkClass}');
+  print('   Chain Id: ${request.chainId}');
+  print('   Log Index: ${request.logIndex}');
+  print('   To: ${request.toAddress}');
+  print('   Token Standard: ${request.tokenStandard}');
+  print('   Amount: ${AmountUtils.addDecimals(request.amount, decimals)}');
```
Comment (KingGorrin) [inline src/bridge.dart]: Token symbol is not displayed

File: src/bridge.dart [hunk 48]
```diff
@@ -0,0 +1,1197 @@
...
+  }
+
+  print('Id: ${request.id}');
+  print('   Network Class: ${request.networkClass}');
+  print('   Chain Id: ${request.chainId}');
+  print('   To: ${request.toAddress}');
+  print(
+      '   From: ${(await znnClient.ledger.getAccountBlockByHash(request.id))?.address}');
+  print('   Token Standard: ${request.tokenStandard}');
+  print('   Amount: ${AmountUtils.addDecimals(request.amount, decimals)}');
```
Comment (KingGorrin) [inline src/bridge.dart]: Token symbol is not displayed

File: src/bridge.dart [hunk 49]
```diff
@@ -0,0 +1,1197 @@
...
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('bridge.guardian.proposeAdmin address');
+    return;
+  }
+
+  String currentAdmin = (await znnClient.embedded.bridge.getBridgeInfo())
+      .administrator
+      .toString();
+  Address newAdmin = Address.parse(args[1]);
```
Comment (KingGorrin) [inline src/bridge.dart]: Check parse and assert not empty or embedded address

File: src/bridge.dart [hunk 50]
```diff
@@ -0,0 +1,1197 @@
...
+  print('Initializing bridge emergency mode ...');
+  AccountBlockTemplate emergency = znnClient.embedded.bridge.emergency();
+  emergency = await znnClient.send(emergency);
+  print('Done');
+}
+
+Future<void> _halt() async {
+  _isAdmin();
+  print('Halting bridge ...');
+  AccountBlockTemplate halt = znnClient.embedded.bridge.halt('');
```
Comment (KingGorrin) [inline src/bridge.dart]: Use signature value '1' to circumvent the empty string unpack issue.

File: src/bridge.dart [hunk 51]
```diff
@@ -0,0 +1,1197 @@
...
+  if (args.length != 3) {
+    print('Incorrect number of arguments. Expected:');
+    print('bridge.admin.revokeUnwrapRequest transactionHash logIndex');
+    return;
+  }
+
+  Hash transactionHash = Hash.parse(args[1]);
+  int logIndex = int.parse(args[2]);
+
+  print('Removing token pair ...');
```
Comment (KingGorrin) [inline src/bridge.dart]: Invalid print statement.

File: src/bridge.dart [hunk 52]
```diff
@@ -0,0 +1,1197 @@
...
+      Address guardian = Address.parse(args[i]);
+      if (guardian != emptyAddress) {
+        guardians.add(guardian);
+      }
+    } catch (e) {
+      print('${red('Error!')} ${args[i]} is not a valid address');
+      return;
+    }
+  }
+
```
Comment (KingGorrin) [inline src/bridge.dart]: Nomination works with time challenges, see C# impl. for details.

File: src/bridge.dart [hunk 53]
```diff
@@ -0,0 +1,1197 @@
...
+Future<void> _changeAdmin() async {
+  _isAdmin();
+
+  if (args.length != 2) {
+    print('Incorrect number of arguments. Expected:');
+    print('bridge.admin.changeAdmin address');
+    return;
+  }
+
+  Address newAdmin = Address.parse(args[1]);
```
Comment (KingGorrin) [inline src/bridge.dart]: Check parse and assert not empty or embedded address.

File: src/bridge.dart [hunk 54]
```diff
@@ -0,0 +1,1197 @@
...
+  print('    bridge.admin.nominateGuardians address1 address2 ... addressN');
+  print('    bridge.admin.changeAdmin address');
+  print('    bridge.admin.setMetadata metadata');
+  print(
+      '    bridge.admin.setOrchestratorInfo windowSize keyGenThreshold confirmationsToFinality estimatedMomentumTime');
+  print(
+      '    bridge.admin.setNetwork networkClass chainId name contractAddress metadata');
+  print('    bridge.admin.removeNetwork networkClass chainId');
+  print('    bridge.admin.setNetworkMetadata networkClass chainId metadata');
+}
```
Comment (KingGorrin) [inline src/bridge.dart]: `bridge.admin.setRedeemDelay` is missing?

**Review by KingGorrin — CHANGES_REQUESTED** [review]

**Review by KingGorrin — CHANGES_REQUESTED** [review]
@Sol-Sanctum I've reviewed the changes and made lots of inline comments. Most are about assertions and inconsistencies between other methods. I'm not sure about all the comments, so if you can give your rationale I'd appreciate it. I've already included your changes in the `bigint` branch of the C# CLI. I await your answers in order to finalize the implementation.

File: src/bridge.dart (hunk 54 above)
Comment (sol-znn) [reply src/bridge.dart]: Omitted on purpose due to incomplete implementation in go-zenon.

1. [Definition statement](https://github.com/zenon-network/go-zenon/blob/master/vm/embedded/definition/bridge.go#L101)
2. [Missing abi](https://github.com/zenon-network/go-zenon/blob/master/vm/embedded/embedded.go#L59)
3. [Missing SetRedeemDelayMethod](https://github.com/zenon-network/go-zenon/blob/master/vm/embedded/implementation/bridge.go)

Comment (KingGorrin) [reply src/bridge.dart]: I should have checked go-zenon. Thanks for the insight.

File: src/general.dart (hunk 3 above)
Comment (sol-znn) [reply src/general.dart]: Agree, it wasn't really valuable information.

Comment (KingGorrin) [reply src/general.dart]: The `process.info` command shows all the info you need.

Also, we need to do something about the versioning schema. I would like to align my versions, but I'm using semantic versioning.

**Comment by sol-znn** [comment]
Thank you for reviewing my work and providing detailed feedback. 
I believe I've addressed most of your comments in this commit: https://github.com/hypercore-one/znn_cli_dart/pull/2/commits/77a3b7fae06f1e713d4c0027ed169d70c4951908

A few stood out, though:
1. [Semantic versioning](https://github.com/hypercore-one/znn_cli_dart/pull/2#discussion_r1263384647)
2. I don't understand [this one](https://github.com/hypercore-one/znn_cli_dart/pull/2#discussion_r1259528074)?
3. I don't think [this](https://github.com/hypercore-one/znn_cli_dart/pull/2#discussion_r1260966259) needs a populated string but we can fix it in a future commit. Will have to check out [this issue.](https://github.com/hypercore-one/znn_sdk_dart/issues/1)

File: lib/accelerator.dart [hunk 55]
```diff
@@ -1,6 +1,6 @@
 import 'package:dcli/dcli.dart' hide verbose;
+import 'package:znn_cli_dart/lib.dart';
 import 'package:znn_sdk_dart/znn_sdk_dart.dart';
-import 'src.dart';
 
 void acceleratorMenu() {
   print('  ${white('Accelerator Z')}');
```
Comment (KingGorrin) [inline lib/accelerator.dart]: Accelerator-Z 🤣

**Review by KingGorrin — APPROVED** [review]

File: src/htlc.dart (hunk 9 above)
Comment (KingGorrin) [reply src/htlc.dart]: Forget this one.

**Comment by KingGorrin** [comment]
Very nice work in total!

**Pull request closed without merging**